File: common/src/main/java/cc/turtl/chiselmon/api/filter/FilterCondition.java
package cc.turtl.chiselmon.api.filter;

import com.cobblemon.mod.common.pokemon.Pokemon;

import java.util.List;
import java.util.function.Predicate;

/**
 * Sealed tree node hierarchy representing a composable filter condition.
 *
 * <p>A filter condition is a tree where:
 * <ul>
 *   <li>{@link And} - ALL children must match (logical AND)</li>
 *   <li>{@link Or}  - ANY child must match (logical OR)</li>
 *   <li>{@link Not} - inverts a single child (logical NOT)</li>
 *   <li>{@link Tag} - a leaf predicate resolved by {@link FilterTagParser}</li>
 * </ul>
 *
 * <p>Conditions are stored as plain strings in {@link FilterDefinition#conditionString}
 * and parsed at runtime by {@link FilterConditionParser}. This type is an intermediate
 * representation used only during compilation to a {@link Predicate}.
 */
public sealed interface FilterCondition permits
        FilterCondition.And,
        FilterCondition.Or,
        FilterCondition.Not,
        FilterCondition.Tag {

    /** All children must match. An empty AND is always true. */
    record And(List<FilterCondition> children) implements FilterCondition {
        public And {
            // Canonical constructor: enforce immutability and null-safety
            children = children != null ? List.copyOf(children) : List.of();
        }
    }

    /** At least one child must match. An empty OR is always false. */
    record Or(List<FilterCondition> children) implements FilterCondition {
        public Or {
            children = children != null ? List.copyOf(children) : List.of();
        }
    }

    /** Inverts a single child condition. */
    record Not(FilterCondition child) implements FilterCondition {}

    /** Leaf node: resolved to a Predicate by {@link FilterTagParser}. */
    record Tag(String tag) implements FilterCondition {
        public Tag {
            tag = tag != null ? tag : "";
        }
    }
    /**
     * Compiles this condition tree into a {@link Predicate}{@code <Pokemon>}.
     *
     * <ul>
     *   <li>AND -  all children must pass; empty AND is always true</li>
     *   <li>OR  - any child must pass; empty OR is always false</li>
     *   <li>NOT - inverts the child predicate</li>
     *   <li>Tag - delegated to {@link FilterTagParser}</li>
     * </ul>
     */
    default Predicate<Pokemon> toPredicate() {
        return switch (this) {
            case Tag tag   -> FilterTagParser.parse(tag.tag());
            case Not not   -> not.child().toPredicate().negate();
            case And and   -> and.children().stream()
                    .map(FilterCondition::toPredicate)
                    .reduce(Predicate::and)
                    .orElse(p -> true);
            case Or or     -> or.children().stream()
                    .map(FilterCondition::toPredicate)
                    .reduce(Predicate::or)
                    .orElse(p -> false);
        };
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/core/api/filter/FilterCondition.kt
package cc.turtl.chiselmon.core.api.filter

import com.cobblemon.mod.common.pokemon.Pokemon
import java.util.function.Predicate

/**
 * Sealed class hierarchy representing a composable filter condition tree.
 *
 * A filter condition is a tree where:
 * - [And] requires ALL children to match (logical AND)
 * - [Or] requires ANY child to match (logical OR)
 * - [Not] inverts a single child (logical NOT)
 * - [Tag] is a leaf predicate resolved by [FilterTagParser]
 *
 * Conditions are stored as plain strings in [FilterDefinition.conditionString]
 * and parsed at runtime by [FilterConditionParser]. This type is an intermediate
 * representation used only during compilation to a [Predicate].
 */
sealed class FilterCondition {

    /** All children must match. An empty AND is always true. */
    data class And(val children: List<FilterCondition>) : FilterCondition()

    /** At least one child must match. An empty OR is always false. */
    data class Or(val children: List<FilterCondition>) : FilterCondition()

    /** Inverts a single child condition. */
    data class Not(val child: FilterCondition) : FilterCondition()

    /** Leaf node: resolved to a Predicate by [FilterTagParser]. */
    data class Tag(val tag: String) : FilterCondition()

    /**
     * Compiles this condition tree into a [Predicate] for [Pokemon].
     *
     * - AND: all children must pass; empty AND is always true
     * - OR: any child must pass; empty OR is always false
     * - NOT: inverts the child predicate
     * - Tag: delegated to [FilterTagParser]
     */
    fun toPredicate(): Predicate<Pokemon> = when (this) {
        is Tag -> FilterTagParser.parse(tag)
        is Not -> child.toPredicate().negate()
        is And -> children
            .map { it.toPredicate() }
            .reduceOrNull { a, b -> a.and(b) }
            ?: Predicate { true }

        is Or -> children
            .map { it.toPredicate() }
            .reduceOrNull { a, b -> a.or(b) }
            ?: Predicate { false }
    }
}